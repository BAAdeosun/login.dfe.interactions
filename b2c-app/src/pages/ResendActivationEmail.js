import React from 'react';
import { animateScroll } from "react-scroll";

import components from '../components';
import { onChange } from '../helpers/pageUpdatesHandler';

class ResendActivationEmail extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            email: null,
            showErrors: false,
            showB2CErrors: true,
            errors: []
        }
        this.handleSubmit = this.handleSubmit.bind(this);
        this.onChange = onChange.bind(this);
    }

    handleSubmit(e) {
        e.preventDefault();
        //hide B2C errors and only show again if we are going to submit the form to B2C
        this.setState({ showB2CErrors: false });
        //update error messages
        this.state.errors.forEach((error) => {
            error.visible.text = error.current.text;
            error.visible.showSummaryText = true;
        });
        //do something to validate and decide if we submit or show errors
        if (this.state.email) {
            //hide our validation errors and prepare to show B2C ones (in case there are any)
            this.setState({ showErrors: false, showB2CErrors: true });
            //everything is valid, set data and submit B2C form
            this.setDataAndSubmit();
        }
        else {
            //show errors in each component
            this.setState({ showErrors: true });
            //scroll to the top of the page to show the errors
            animateScroll.scrollToTop({ duration: 500 });
        }
    }

    setDataAndSubmit() {
        //retrieve all elements we will need and set their values
        let b2cEmail = document.getElementById('email');
        let b2cSubmitButton = document.getElementById('continue');

        if (b2cEmail && b2cSubmitButton) {
            b2cEmail.value = this.state.email;
            //submit B2C form
            b2cSubmitButton.click();
        }
    }

    render() {

        const formContent =
            <div>
                <components.InputField
                    type='email'
                    inputId='email'
                    inputLabel='Email address'
                    onChange={this.onChange}
                    errorMessagePlaceholder='email address'
                    showErrors={this.state.showErrors}
                    errors={this.state.errors} />
            </div>

        /**
         * Page configuration
         */
        const title = 'Resend activation email'
        const pageConfig = {
            title: title,
            header: title,
            formContent: formContent,
            submitButtonText: 'Send activation email',
            submitHandler: this.handleSubmit,
            errors: this.state.errors,
            showB2CErrors: this.state.showB2CErrors
        };


        return (
            <div id="resendActivationEmail">
                <components.PageContainer pageConfig={pageConfig} />
            </div>
        )
    }
}

export default ResendActivationEmail;
