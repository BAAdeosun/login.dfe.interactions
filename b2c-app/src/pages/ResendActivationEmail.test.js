
import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { JSDOM } from "jsdom";

import ResendActivationEmail from './ResendActivationEmail';

it('renders without crashing', () => {
    shallow(<ResendActivationEmail />);
});

it('renders correctly', () => {
    const tree = renderer
        .create(<ResendActivationEmail />)
        .toJSON();
    expect(tree).toMatchSnapshot();
});

describe('when submitting the form', () => {

    describe('when all details have not been set', () => {

        it('sets showErrors to true and does not call submit function', () => {
            const wrapper = mount(<ResendActivationEmail />);
            const form = wrapper.find('#customForm');
            const setDataAndSubmitSpy = jest.spyOn(wrapper.instance(), 'setDataAndSubmit');

            form.simulate('submit');

            expect(wrapper.state().showErrors).toBe(true);
            expect(setDataAndSubmitSpy).not.toHaveBeenCalled();

        });

    });

    describe('when all details have been set', () => {

        let wrapper;
        let form;
        let setDataAndSubmitSpy;
        let b2cSubmitSpy;

        beforeEach(() => {
            //mount page with values in state for all required fields
            wrapper = mount(<ResendActivationEmail />);

            wrapper.setState({
                email: 'a'
            });

            //get form element to simulate submit in tests
            form = wrapper.find('#customForm');

            //spy on the page submit button
            setDataAndSubmitSpy = jest.spyOn(wrapper.instance(), 'setDataAndSubmit');

            //create some DOM elements to simulate the content added by B2C
            const dom = new JSDOM();
            global.document = dom.window.document;

            //add the B2C submit button at least and spy on its click event
            let b2cSubmitElem = global.document.createElement('button');
            b2cSubmitElem.id = 'continue';
            global.document.body.appendChild(b2cSubmitElem);

            //spy on b2c submit button click
            b2cSubmitSpy = jest.spyOn(global.document.getElementById('continue'), 'click');
        });

        describe('when not all B2C DOM elements are present', () => {

            it('sets showErrors to false and calls submit function but does not call B2C submit', () => {

                form.simulate('submit');

                expect(setDataAndSubmitSpy).toHaveBeenCalled();
                expect(wrapper.state().showErrors).toBe(false);
                expect(b2cSubmitSpy).not.toHaveBeenCalled();
            });
        })

        describe('when all B2C DOM elements are present', () => {

            it('sets showErrors to false and calls submit function and then the B2C submit button click', () => {

                //add B2C content to the DOM
                let b2cElem = global.document.createElement('div');
                b2cElem.innerHTML = `
                    <span id='resend_email'></span>
                `;
                global.document.body.appendChild(b2cElem);

                form.simulate('submit');

                expect(setDataAndSubmitSpy).toHaveBeenCalled();
                expect(wrapper.state().showErrors).toBe(false);
                expect(b2cSubmitSpy).toHaveBeenCalled();
            });

        });

    });

});
