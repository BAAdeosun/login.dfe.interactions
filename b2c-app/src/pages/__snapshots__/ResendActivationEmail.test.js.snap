// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  id="resendActivationEmail"
>
  <div>
    <div
      className="govuk-width-container"
    >
      <div
        className="govuk-breadcrumbs"
      >
        <ol
          className="govuk-breadcrumbs__list"
        >
          <li
            className="govuk-breadcrumbs__list-item"
          >
            <a
              className="govuk-breadcrumbs__link"
              href="https://nationalcareers.service.gov.uk/"
            >
              Home
            </a>
          </li>
        </ol>
      </div>
      <main
        className="govuk-main-wrapper"
      >
        <div
          className="govuk-grid-row"
        >
          <div
            className="govuk-grid-column-two-thirds"
          >
            <div
              className="pageLevelErrorContainer hide"
            >
              <div
                aria-labelledby="error-summary-title"
                className="govuk-error-summary"
                data-module="govuk-error-summary"
                role="alert"
                tabIndex="-1"
              >
                <h2
                  className="govuk-error-summary__title"
                  id="error-summary-title"
                >
                  There is a problem
                </h2>
                <div
                  className="govuk-error-summary__body"
                >
                  <ul
                    className="govuk-list govuk-error-summary__list"
                    id="errorSummaryItems"
                  />
                </div>
              </div>
            </div>
            <h1
              className="govuk-heading-xl govuk-!-margin-top-6 govuk-!-margin-bottom-9"
            >
              Resend activation email
            </h1>
            <div>
              <p
                className="govuk-body"
              >
                Enter the email you used to create your account. We'll send you a new account activation email.
              </p>
            </div>
            <form
              id="customForm"
              noValidate={true}
              onSubmit={[Function]}
            >
              <div>
                <div
                  className="govuk-form-group "
                >
                  <label
                    className="govuk-label"
                    htmlFor="emailCustom"
                  >
                    Email address
                  </label>
                  <input
                    autoComplete="email"
                    className="govuk-input govuk-!-width-one-half"
                    id="emailCustom"
                    name="email"
                    noValidate={true}
                    onChange={[Function]}
                    spellCheck="false"
                    type="email"
                  />
                </div>
              </div>
              <button
                className="govuk-button"
                id="preSubmit"
                type="submit"
              >
                Send activation email
              </button>
            </form>
          </div>
        </div>
      </main>
    </div>
  </div>
</div>
`;
